### ISSUE TITLE
Review and suggestions of the returned data (schema)

### ISSUE THREAD

**Issue by emagnier** [romcal/romcal#180]
In this issue I propose to review the actual returned data by romcal, and suggest some improvements.

TL;DR: check the example under the "Proposed returned data" part.

It might be not too difficult to adjust the schema for existing features/data. And this could be a good opportunity to do it now with the codebase just upgraded to TS.

@pejulian and @tukusejssirs, let me know your thoughts.

---

## Actual data :

Here the actual returned data (from the `_F-TypeScript` branch):
```json
{
  "key": "fridayOfTheFirstWeekOfLent",
  "name": "Friday of the 1st week of Lent",
  "date": "2020-03-06T00:00:00.000Z",
  "type": 8,
  "data": {
    "season": {
      "key": "LENT",
      "value": "Lent"
    },
    "meta": {
      "titles": [],
      "psalterWeek": {
        "key": 4,
        "value": "Easter"
      },
      "liturgicalColor": {
        "key": "PURPLE",
        "value": "[object Object]"
      },
      "cycle": {
        "key": 0,
        "value": "Year A"
      }
    },
    "calendar": {
      "weeks": 53,
      "week": 10,
      "day": 65
    }
  },
  "_id": 834,
  "_stack": 0
}
```

A few comments on some properties:

#### → `type`

Should be a `string` containing the key name of the rank of this celebration. Maybe it's also a good opportunity to rename it to `rank` (so it will be closer to the vocabulary used in the _Universal Norms on the Liturgical Year and the General Roman Calendar_).

#### → `_id` and `_stack`

Ideally, they should not be output. They are used internally and should not be used outside of romcal.

#### → `data/season`

Instead we might return an array:
```
[{ key: `LENT`, value: 'Lent' }]
```
Because some seasons can overlap. This is for example the case for Lent, the Easter Triduum and Estertide: 
 - Lent end on the Holy Thursday just before the evening mass (_In Cena Domini_)
 - The Easter Triduum start from this mass, until Easter Sunday’s Vespers.
 - Eastertide starts the Sunday of Ester (included).

Official liturgical seasons are `advent`, `christmastide`, `ordinary-time`, `lent`, `easter-triduum`, `eastertide`.

In addition to `seasons`, we might consider offering another property for all kinds of periods. This `periods` property should also be an array, and might contain:
- Official liturgical periods and octaves: `christmas-octave`, `holy-week`, `easter-octave`.
- Other periods, for convenient usage: `early-ordinary-time`, `later-ordinary-time`.

#### → `data/calendar`

In addition to the existing properties, some new property will be useful to get contextual data, especially for monastic usages:
- `dow` (number) (or `dayOfWeek`) current day of week, 0 is Sunday, 6 is Saturday
- 'weekInSeason` (number): the current week in a specific season. For example `1` for `thursdayOfTheFirstWeekOfLent`, `12` for `saturdayOfTheTwelfthWeekOfOrdinaryTime`, `0` for `ashWednesday`, `1` for `christmas`...

Additionally, we could rename the existing properties to make them more explicit.
- `weeks` to `totalWeeksInYear` (furthermore is it civil or liturgical year?)
- `week` to `weekInYear`
- `day` to `dayInYear`

#### → `data/meta/liturgicalColor`

There is already a discussion about that here: #177

#### → `data/meta/psalterWeek`

I'm wondering if `key` should start from `1` instead of `0`, so `1` correspond to `Week I`. Also to be consistent with other metadata, should it return a string instead, for example `WEEK_1`? The `value` should be a localised string, actually this is not the case.

#### → `data/meta/cycle`

Should be renamed to `sundayCycle`. Again here I'm wondering if the key should be a string containing something like `YEAR_A`. The `value` should be a localised string, actually this is not the case.

Additionally, we should consider to add a `feriaCycle` to get data one `YEAR_1` / `YEAR_2` (or `YEAR_ODD` / `YEAR_EVEN`).

#### → `data` and `data/meta` properties

They are all data and metadata, so I'm wondering if everything inside `meta` could be at the same level of everything in the `data` object. Actually (at least for me), it's not very clear why some data are in `data` and some other are in `meta`. I suggest removing `meta` and move everything in `data`. And finally, for a reading perspective, I propose to rename `data` to `meta` (or `metadata`) to make this property name a little bit more different than the `date` property (I find the `data` and `date` names too similar).

The `prioritized` property has the same importance as the `type` property when romcal compute and organise celebrations for a specific day. This is in fact not a liturgical information, but a convenient trick to play with ranks and celebration precedences. So instead to be in the `data` object property, I suggest moving it to the root object (at the same level of `type`).

## Proposed returned data:

```jsonc
{
  "key": "fridayOfTheFirstWeekOfLent",
  "name": "Friday of the 1st week of Lent",
  "date": "2020-03-06T00:00:00.000Z",
  "rank": "FERIA",
  "prioritized": false,
  "metadata": {
    "season": [
      {
        "key": "LENT",
        "value": "Lent"
      }
    ],
    "titles": [],
    "psalterWeek": {
      "key": "WEEK_1",
      "value": "Psalms week 1"
    },
    "liturgicalColor": {
      "key": "PURPLE",
      "value": "Purple"
    },
    "sundayCycle": {
      "key": "YEAR_A",
      "value": "Year A"
    },
    "feriaCycle": {
      "key": "YEAR_1",
      "value": "Year 1"
    },
    "calendar": {
      "totalWeeksInYear": 53,
      "weekInYear": 10,
      "dayInYear": 65,
      "dayOfWeek": 5,
      "weekInSeason": 1
    },
    "fromCalendar": "temporal", 
  },
  "base": null // Or should not be output, because the current celebration is already a Feria
}
```

You will notice 2 others proposed properties:

#### → `metadata/fromCalendar`

Represent the name from where this celebration is defined. E.g. "slovakia".
Or "temporal" when it comes from the "Seasons.ts" ; or "general" when it comes from "Dates.ts" or "calendars/general.ts".

#### → `metadata/fromCombinedCalendars`

An array that contains the name of all the combined calendar. Actually, it could be one of them:
```
[ 'temporal' ]
[ 'temporal', 'general' ]
[ 'temporal', 'general', 'slovakia' ]
```
Then, when calendar inheritance support will be added (#48), this could look like:
```
[ 'temporal', 'general', 'france', 'paris', 'local' ]
```

The `fromCalendar` and `fromCombinedCalendars` properties will provide useful information to know from where a celebration is defined, and the final data for a day/celebration conputed from a combination  of calendars.

#### → Other possible properties

- `metadata/temporal` or `metadata/sanctoral` (or another generic name) that could include specific information from a future temporal/sanctoral db.
- `metadata/masses` that contain readings (or at least references), prayers...

**Comment by tukusejssirs** [comment]
> ```json
> ...
> "type": 8,
> ...
> "key": 4,
> ...
> ```

What are these keys, what do there mean? Could we not use some self-explanatory strings instead of the numbers?

---

@emagnier, I had the same thoughts after reading the JSON. :smiley: :+1:

> `dow` (number) (or `dayOfWeek`) current day of week, 0 is Sunday, 6 is Saturday

I vote for `dayOfWeek` as it is more human-readable. :smiley:

> - `weekInSeason` (number): the current week in a specific season. For example `1` for `thursdayOfTheFirstWeekOfLent`, `12` for `saturdayOfTheTwelfthWeekOfOrdinaryTime`, `0` for `ashWednesday`, `1` for `christmas`...

And what about `dayInSeason`? :smiley:


> - `weeks` to `totalWeeksInYear` (furthermore is it civil or liturgical year?)
> - `week` to `weekInYear`
> - `day` to `dayInYear`

Because of the ambiguity in the `year`, I would suggest to following:
- `weeks` to `totalWeeksInCivilYear`;
- `week` to `weekInCivilYear`;
- `day` to `dayInCivilYear`.

> #### → `data/meta/psalterWeek`
> 
> I'm wondering if `key` should start from `1` instead of `0`, so `1` correspond to `Week I`. Also to be consistent with other metadata, should it return a string instead, for example `WEEK_1`? The `value` should be a localised string, actually this is not the case.

Why would you think it should be string? If the key is named `psalterWeek`, it is self-explanatory, and thus the `Week` (or `WEEK`) would be kind of a duplicity. I’d suggest to make it an integer value between `1..4`.

> #### → `data/meta/cycle`
> 
> Should be renamed to `sundayCycle`. Again here I'm wondering if the key should be a string containing something like `YEAR_A`. The `value` should be a localised string, actually this is not the case.

I agree with the change of the key name to `sundayCycle`. But I would suggest to let the value be a single letter from `[ABC]`. I think that the localisation here is not necessary.

> Additionally, we should consider to add a `feriaCycle` to get data one `YEAR_1` / `YEAR_2` (or `YEAR_ODD` / `YEAR_EVEN`).

I agree with the creation of it, but it should be named `ferialCycle` (note the letter `l`). And I suggest that the key value should be an integer (`1..2`) and no need for localisation then. KISS is my motto, brothers. :smiley:

> #### → `data` and `data/meta` properties
> 
> They are all data and metadata, so I'm wondering if everything inside `meta` could be at the same level of everything in the `data` object. Actually (at least for me), it's not very clear why some data are in `data` and some other are in `meta`. I suggest removing `meta` and move everything in `data`. And finally, for a reading perspective, I propose to rename `data` to `meta` (or `metadata`) to make this property name a little bit more different than the `date` property (I find the `data` and `date` names too similar).

I agree that the placement of the keys in either `data` or `data/meta` is quite ambiguous. On the other hand, I think that some division should be made:
- celebration-specific data (like name, saint/blessed titles, liturgical colour);
- season-related data (like season name, day in season, week in season);
- breviary-related data (for now, only the psalter week);
- Mass-related data (for the future usage);
- liturgical-year-related data (Sunday and ferial cycles);
- calendar-related data [like calendar name (e.g. `cal1962`), calendar ‘type’ (like `general` or `particular`), `fromCalendar` (like `england` or `slovakia`)].

> ```json
> "season": [
>   {
>     "key": "LENT",
>     "value": "Lent"
>   }
> ],
> ```

May I know why we have two keys (`key` and `value`) with essentially the same values (`LENT` and `Lent`)? I think it is a duplicity.

I think that `season` could contain both string and array values, that is: when there is only one season (which is usually the case), it should contain a string (`"season": "Lent"`); when there is more then one season, we should use an array:

```json
"season": [
  "Ordinary Time",
  "Early Ordinary Time"
]
```

Same would apply to `liturgicalColor`, `sundayCycle`, `ferialCycle`.

> Then, when calendar inheritance support will be added (#48), this could look like:
> 
> ```
> [ 'temporal', 'general', 'france', 'paris', 'local' ]
> ```

I like it very much. But I think we should add the continent too. Why? Because some of the calendar definitions should be made in the continent-specific celebrations (for the `Patron of Europe` and the like).

---

With the rest (stuff I have not commented on) I agree with @emagnier.

**Comment by emagnier** [comment]
> And what about dayInSeason?

I don't see a use case for that, but why not.

> Why would you think it should be string? If the key is named psalterWeek, it is self-explanatory, and thus the Week (or WEEK) would be kind of a duplicity. I’d suggest to make it an integer value between 1..4.

It's already self-explanatory because it's nested in a `psalterWeek` object: `psalterWeek.key` and `psalterWeek.value`.

> I agree with the change of the key name to sundayCycle. But I would suggest to let the value be a single letter from [ABC]. I think that the localisation here is not necessary.

Why isn't the localisation necessary? The localisation might not be necessary for everyone (it's why there is a `key` property). But also having a localised `value` will avoid to do translation specifically for that when we need it.

> I agree with the creation of it, but it should be named ferialCycle (note the letter l). 

I personally prefer `feriaCycle`: using `feria` as a noun. Here it's not a literary style but a coding style, where `feria` could be considered as a namespace, and `cycle` one of its properties.

> May I know why we have two keys (key and value) with essentially the same values (LENT and Lent)? I think it is a duplicity.

Because `key` contain a constant which will never change. `value` is the localised equivalent. Here the example is in English so they are quite similar, but for example in French the `value` will contain `Carême`.

> I think that season could contain both string and array values.

I will discourage that. One property should always have a unique type, not multiple possible types. This will avoid doing type checking when we consume data from romcal, and could also avoid potential bugs/issues.

Also we still need the `{ key, value }` object. Because depending on our needs we could lean on a `key` containing a constant that will never change, or a localised `value` for display usage.

> I like it very much. But I think we should add the continent too. 

This is only an example. With the calendar inheritance architecture, we will be able to organise the calendar stacking as we want.

---

For other points, I will take time to think about it, and propose a revised data model (probably next Monday or Tuesday)

**Comment by tukusejssirs** [comment]
I was thinking about this issue while working on the separation of multiple celebrations.

It came to my mind that if we want to make the JSON response as agnostic as possible, should we not send all the metadata separately?

What I mean is that currently, we send the _formatted_ celebration name as a string. I suggest that besides the formatted celebration name, we should send all the metadata the celebration name is composed of, also separately.

By the metadata I mean the following information (I suggest to create `personMetadata[]` array for persons and `eventMetadata[]` for celebration where necessary):
- `canonizationLevel` (either `saint` or `blessed`);
- `numberOfPersons` (how many persons are in the particular celebration; `0` for non-person celebrations);
- `names` (if the celebration is a person celebration, we would output an array of names for the celebrated persons);
- `surnames` (if the celebration is a person celebration, we would output an array of surnames for the celebrated persons);
- `sex` (sex of the persons);
- etc (for more metadata types, see [this sheet](https://docs.google.com/spreadsheets/d/1c9YWtfntpxy0r0rg0mfowJ2v-yKW6wSMmNzyY3umBU4/edit#gid=631058184)).

Reason? If the user would have the separate metadata:
1. we would not have to provide a way to format the celebration name;
2. the user would be able to format the celebration name as they would want.

**Comment by emagnier** [comment]
This is what I proposed to include under these properties:

> `metadata/temporal` or `metadata/sanctoral` (or another generic name) that could include specific information from a future temporal/sanctoral db.

But I had not detailed them (because it's not something implemented yet)

**Comment by emagnier** [comment]
Here an updated proposal of the returned schema.
Let me know your thoughts!

```ts
interface DateItem {
	key: string; // Uniq key name E.g. 'fridayOfTheFirstWeekOfLent'
	name: string; // Localised name. E.g. 'Friday of the 1st week of Lent'
	date: Date; // E.g. '2020-03-06T00:00:00.000Z'
	rank: 'SOLEMNITY' | 'SUNDAY' | 'TRIDUUM' | 'HOLY_WEEK' | 'FEAST' | 'MEMORIAL' | 'OPT_MEMORIAL' | 'COMMEMORATION' | 'FERIA';
	prioritized?: boolean; // Default: false (should be output only if true)
	base: null | DateItem; // Contain the base Feria or Sunday DateItem of the day. Is null if the current celebration is already a Feria or a Sunday

	metadata: {

		seasons: string[]; // [ 'EASTER_TRIDUUM', 'EASTERTIDE' ], or [ 'ORDINARY_TIME' ]
		seasonNames: string[]; // [ 'Easter Triduum', 'Eastertide' ], or [ 'Ordinary Time' ]

		cycles: {
			psalterWeek: 1 | 2 | 3 | 4;
			psalterWeekName: string;  // E.g. 'Psalter week 4'
			sundayCycle: 'A' | 'B' | 'C';
			sundayCycleName: string; // E.g. 'Year A'
			feriaCycle: 1 | 2;
			feriaCycleName: string; // E.g. 'Year II'
		}

		calendar: {
			totalWeeksInCivilYear: number;
			weekInCivilYear: number;
			weekInLiturgicalYear: number;
			weekInSeason: number;
			dayInCivilYear: number;
			dayInLiturgicalYear: number;
			dayInSeason: number;
			dayOfWeek: 0 | 1 | 2 | 3 | 4 | 5 | 6; // 0 is Sunday, 6 is Saturday
		}

		fromCalendar: string; // E.g. 'temporal', or 'general', or 'slovakia'..,
		fromCombinedCalendars: string[]; // E.g. [ 'temporal', 'general', 'france', 'paris', 'local' ]

		celebration: {
			liturgicalColor: 'WHITE' | 'GREEN' | 'RED' | 'PURPLE' | 'ROSE';
			liturgicalColorName: string; // Localised name. E.g. in English 'White', 'Green', 'Red'...
			titles: string[]; // E.g. [ 'DOCTOR_OF_THE_CHURCH' ]
			titlesNames: string[]; // E.g. [ 'Doctor of the Church' ]
			rankName: string; // E.g. 'Optional Memorial'

			// Except for liturgicalColor and titles, content below is TBD
			// the keys here might come from a future sanctoral and temporal db
			// and are given here as example
			type: 'temporal' | 'sanctoral';
			canonisationLevel?: 'Saint' | 'Blessed';
			firstName?: string;
			sex?: 'male' | 'female';
			number?: number;
		}

		masses: Mass[], // Array of mass, TBD (because some celebrations could have multiple masses, e.g. during Vigil)
	}
}
```

(I didn't want to divise data to much in nested objects)

**Discussion**:
1. `metadata`? or `meta`? or should we remove it and move all the content at the root of the `DateItem` object?
2. `psalterWeekName`, `sundayCycleName`, `feriaCycleName` might not be required. Do we want to keep them? On my side I vote yes.
3. `feriaCycle` or `ferialCycle`? On my side I vote for `feriaCycle` for the reasons I've gave [above](https://github.com/romcal/romcal/issues/180#issuecomment-596004001).
4. Anything else about property name, nesting?

**Comment by tukusejssirs** [comment]
I think that the values titles, celebration ranks and canonisation level should be in lowercase. Those words are not proper names (at least not in English).

---

>     1. `metadata`? or `meta`? or should we remove it and move all the content at the root of the `DateItem` object?

I don’t care if it’s `metadata` or `meta`, pick one. :smiley:

>     2. `psalterWeekName`, `sundayCycleName`, `feriaCycleName` might not be required. Do we want to keep them? On my side I vote yes.

We need to decide how much of the localisation we do in romcal. If we want to be minimalistic (in l10n _and_ in the code), then drop them. If we want to have localised as much as possible, then keep them.

On my part, the basic localisation is enough, because (1) the more localisation, the the more flaws; (2) there should be external (client) app that would localise/format the data gathered/computed by romcal.

>     3. `feriaCycle` or `ferialCycle`? On my side I vote for `feriaCycle` for the reasons I've gave [above](https://github.com/romcal/romcal/issues/180#issuecomment-596004001).

I’d vote for `ferialCycle` (just because of English grammer). However, I am not opposed to use _feria_ in the key name, but then I think it is better to change the order of the words: `cycleSunday`, `cycleFeria`.

>     4. Anything else about property name, nesting?

I am wondering if there is such a term as `temporal` in the liturgy. I know that we separate some of the celebrations are defined in the `src/lib/Celebrations.ts` instead of `src/calendars/general.ts`. I think the two files should be merged, because the celebrations in `temporal`/`Celebrations.ts` are part of the General Roman Calendar, even though they cannot be overriden. Even if the two files won’t be merged, it might be a good idea to set the value of `fromCalendar` and `fromCombinedCalendars` to either `general` or both `general` and `temporal`.

**Comment by emagnier** [comment]
About your 4th point: in the liturgy their complete name is temporal cycle and sanctoral cycle.

The **temporal** (or the proper of seasons) is the essential and main part of the liturgical year (mainly sundays and ferial days). It's a cycle that traces the earthly life and the mystery of the Christ, and is deeply related to the liturgical seasons continuity.

They are also gathered in 2 books or 2 lectionaries: the lectionary of sundays (that also include solemnities that can occur on any week days); and the weekday lectionary.

In romcal, this cycle of celebrations is primarily computed in `lib/Calendar.ts`, and some of them in `lib/Dates.js`.

By opposition, the **sanctoral** (or the proper of saints), is an annual cycle of feasts, which consist of fixed dates (on the civil year), not related to seasons. It mainly contains the celebrations of the saints, and some other celebrations for the mystery of the church.

They are gathered in 1 lectionary: the sanctoral book.

In romcal, some celebrations from the sanctoral cycle are computed in `lib/Dates.js` (they are part of the general Roman calendar), and the majority are defined in `src/calendars/*` (including `general.ts` to complete the general Roman calendar, and many particular proper calendars).

Having a separated `lib/Dates.js` library gives us a handy manner to compute dates from each other (giving easy access to the main feasts and solemnity in the codebase).

**Comment by emagnier** [comment]
FYI a new proposition has been made in the `_F-TypeScript`.
Please have a look to the corresponding documentation: https://github.com/romcal/romcal/blob/_F-TypeScript/docs/data-output.md#json-structure

**Comment by emagnier** [comment]
Addressed on https://github.com/romcal/romcal/pull/250.
